gpytorch/likelihoods/hadamard_gaussian_likelihood.py

```diff
@@ -81,17 +83,27 @@ def raw_noise(self, value: torch.Tensor) -> None:
         self.noise_covar.initialize(raw_noise=value)
 
     def _shaped_noise_covar(self, base_shape: torch.Size, *params: Any, **kwargs: Any) -> LinearOperator:
-        # params contains task indexes, shape (*task_batch_shape, num_data, 1)
-        if not params or not params[0]:
+        # params contains input data, shape (*task_batch_shape, num_data, d)
+        if len(params) == 0 or len(params[0]) == 0:
             raise ValueError("Task indices must be provided.")
-        task_idcs = params[0][-1]
+        task_idcs = params[0] if torch.is_tensor(params[0]) else params[0][0]
+        if task_idcs.shape[-1] > 1:
+            if self.task_feature_index is None:
+                raise ValueError("Task indices must be a single dimension if task_feature_index is not provided.")
+            # handle case where full inputs are passed in,
+            # rather than just the task indices
+            task_idcs = task_idcs[..., self.task_feature_index].unsqueeze(-1)
+        task_idcs = task_idcs.long()
```

Comment:
curious why this is needed (before there was no casting to `long`)

Reply:
Previously this assumed that task_idcs (a long-value tensor) was passed in as params[0], but this is adding support for extracting task_idcs from the full train inputs, so we need to cast it to `long`.